File: src/main/java/sage/domain/service/ListService.java
package sage.domain.service;

import java.util.ArrayList;

import httl.util.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.Assert;
import sage.domain.repository.*;
import sage.entity.FollowListEntity;
import sage.entity.ResourceListEntity;
import sage.transfer.*;

@Service
@Transactional
public class ListService {
  @Autowired
  private ResourceListRepository resourceListRepo;
  @Autowired
  private FollowListRepository followListRepo;
  @Autowired
  private FollowRepository followRepo;
  @Autowired
  private UserRepository userRepo;
  @Autowired
  private TagRepository tagRepo;
  
  public ResourceList getResourceList(Long id) {
    return ResourceList.fromEntity(resourceListRepo.nonNull(id));
  }
  
  public Long addResourceList(ResourceList rc, Long userId) {
    ResourceListEntity entity = escaped(rc).toEntity();
    entity.setOwnerId(userId);
    resourceListRepo.save(entity);
    return entity.getId();
  }
  
  public void updateResourceList(ResourceList rc, Long userId) {
    ResourceListEntity entity = resourceListRepo.nonNull(rc.getId());
    Assert.isTrue(entity.getOwnerId().equals(userId));

    ResourceListEntity neo = escaped(rc).toEntity();
    entity.setName(neo.getName());
    entity.setListJson(neo.getListJson());
    resourceListRepo.update(entity);
  }
  
  public FollowList getFollowList(Long id) {
    return FollowListLite.fromEntity(followListRepo.nonNull(id))
        .toFull($ -> new UserLabel(userRepo.nonNull($)), $ -> new TagLabel(tagRepo.nonNull($)));
  }
  
  public Long addFollowList(FollowListLite fcLite, Long ownerId) {
    FollowListEntity entity = fcLite.toEntity();
    entity.setOwnerId(ownerId);
    followListRepo.save(entity);
    return entity.getId();
  }
  
  public void updateFollowList(FollowListLite fcLite, Long userId) {
    FollowListEntity entity = followListRepo.nonNull(fcLite.getId());
    Assert.isTrue(entity.getOwnerId().equals(userId));

    FollowListEntity neo = fcLite.toEntity();
    entity.setName(neo.getName());
    entity.setListJson(neo.getListJson());
    followListRepo.update(entity);
  }

  private ResourceList escaped(ResourceList rc) {
    ResourceList neo = new ResourceList(rc.getId(), rc.getOwnerId(), rc.getName(), new ArrayList<>());
    rc.getList().forEach(info -> neo.getList().add(
        new ResourceInfo(StringUtils.escapeXml(info.getLink()), StringUtils.escapeXml(info.getDesc()))));
    return neo;
  }
}

File: src/main/kotlin/sage/service/ListService.kt
package sage.service

import httl.util.StringUtils
import org.springframework.stereotype.Service
import org.springframework.util.Assert
import sage.entity.FollowListEntity
import sage.entity.ResourceListEntity
import sage.entity.Tag
import sage.entity.User
import sage.transfer.*
import sage.util.Strings
import java.util.*

@Service
class ListService {

  fun getResourceList(id: Long): ResourceList {
    return ResourceList.fromEntity(ResourceListEntity.get(id))
  }

  fun addResourceList(rc: ResourceList, userId: Long): Long? {
    val entity = escaped(rc).toEntity()
    entity.ownerId = userId
    entity.save()
    return entity.id
  }

  fun updateResourceList(rc: ResourceList, userId: Long) {
    val entity = ResourceListEntity.get(rc.id)
    Assert.isTrue(entity.ownerId == userId)

    val neo = escaped(rc).toEntity()
    entity.name = neo.name
    entity.listJson = neo.listJson
    entity.update()
  }

  fun getFollowList(id: Long): FollowList {
    return FollowListLite.fromEntity(FollowListEntity.get(id)).toFull({UserLabel(User.get(it)) }) { TagLabel(Tag.get(it)) }
  }

  fun addFollowList(fcLite: FollowListLite, userId: Long): Long {
    val entity = fcLite.toEntity()
    entity.ownerId = userId
    entity.save()
    return entity.id
  }

  fun updateFollowList(fcLite: FollowListLite, userId: Long) {
    val entity = FollowListEntity.get(fcLite.id)
    Assert.isTrue(entity.ownerId == userId)

    val neo = fcLite.toEntity()
    entity.name = neo.name
    entity.listJson = neo.listJson
    entity.update()
  }

  private fun escaped(rc: ResourceList): ResourceList {
    val neo = ResourceList(rc.id, rc.ownerId, rc.name, ArrayList<ResourceInfo>())
    rc.list.forEach { info ->
      neo.list.add(ResourceInfo(Strings.escapeHtmlTag(info.link), Strings.escapeHtmlTag(info.desc)))
    }
    return neo
  }

  fun followListsOfOwner(ownerId: Long): List<FollowList> {
    return FollowListEntity.byOwner(ownerId).map { l ->
      FollowListLite.fromEntity(l).toFull({ UserLabel(User.get(it)) }, { TagLabel(Tag.get(it)) })
    }
  }

  fun resourceListsOfOwner(ownerId: Long): List<ResourceList> {
    return ResourceListEntity.byOwner(ownerId).map { l -> ResourceList.fromEntity(l) }
  }
}
